### PR TITLE
Add new sharing user

### PR DESCRIPTION
**Pull request by barrymcgee**
## Done

Add new sharing user to model details

## QA

- Pull code
- Run `./run clean && ./run serve`
- Open http://0.0.0.0:8036/
- Open a model
- Click the 'Share model' button in the InfoPanel
- Check that you can add/remove users

## Details

Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1744,
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1742,
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1738.
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1740,
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1745

### CODE REVIEW

**Comment by barrymcgee** [comment]
@hatched Testing issue has been added to this epic as discussed https://github.com/canonical-web-and-design/juju-squad/issues/1770

File: src/components/Aside/Aside.tsx [hunk 1]
```diff
@@ -25,6 +27,7 @@ export default function Aside({
         "is-narrow": width === "narrow",
         "is-wide": width === "wide",
         "is-pinned": pinned === true,
+        "is-split": isSplit === true,
```
Comment (hatched) [inline src/components/Aside/Aside.tsx:30]: Why is this necessary?

File: src/components/Aside/_aside.scss [hunk 2]
```diff
@@ -19,4 +19,41 @@
     justify-content: center;
     width: 100%;
   }
+
+  &.is-split {
```
Comment (hatched) [inline src/components/Aside/_aside.scss:23]: This split logic seems like it's very dependent on the content and should be part of the sharing styles and not the aside itself.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 3]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
+                              modelControllerURL,
+                              modelUUID,
+                              store.getState,
+                              updatedUserAccess,
+                              usersAccess?.[userName],
+                              "grant",
+                              dispatch
+                            );
+                            const error = response?.results[0]?.error?.message;
+                            if (error) {
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:210]: Is it possible to error now that you have the UI setup correctly?

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 4]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
             </div>
           }
         />
-        <div className="p-panel__content">
-          <div>
+        {errorMsg && (
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:127]: We don't have a design for this. How can we make it fail to qa?

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 5]
```diff
@@ -1,22 +1,89 @@
...
+
+    users?.forEach((user: user) => {
+      const displayName = user["user"];
+
+      if (clonedUserAccess) {
+        clonedUserAccess[displayName] = user?.["access"];
+        setUsersAccess(clonedUserAccess);
+      }
+    });
+  }, [users]); // eslint-disable-line
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:85]: typically you should put the specific rule you're ignoring so that it'll still check the others.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 6]
```diff
@@ -28,8 +95,26 @@ export default function ShareModel() {
     access: string;
   };
 
+  const userAlreadyHasAccess = (userName: string) => {
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:98]: This should take the `users` variable as an argument to make it pure.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 7]
```diff
@@ -28,8 +95,26 @@ export default function ShareModel() {
...
+    return users.some((userEntry: User) => userEntry.user === userName);
+  };
+
+  const handleRemoveUser = async (userName: string) => {
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:102]: Can you add the arguments to this as well. That way the function can be called without having to worry about the context that the function was declared in vs when it was called.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 8]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
-                    {userObj["last-connection"]
-                      ? formatFriendlyDateToNow(userObj["last-connection"])
+                    {lastConnected
+                      ? formatFriendlyDateToNow(lastConnected)
                       : `Never connected`}
-                  </p>
+                    <Formik
+                      initialValues={{
+                        selectAll: false,
+                        selectedUnits: [],
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:180]: I'm not sure what any of these props are for?

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 9]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
+                      }}
+                      onSubmit={() => {}}
+                    >
+                      <FormikFormData>
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:184]: You can just use `Form` here instead.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 10]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
+                        <Field
+                          as="select"
+                          name="accessLevel"
+                          onChange={async (
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:188]: Can you move this out of the markup plz.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 11]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
+                            e: React.ChangeEvent<HTMLInputElement>
+                          ) => {
+                            const cloneUserAccess = cloneDeep(usersAccess);
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:191]:
```suggestion
                            const clonedUserAccess = cloneDeep(usersAccess);
```

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 12]
```diff
@@ -28,8 +95,26 @@ export default function ShareModel() {
...
   };
 
+  const userAlreadyHasAccess = (userName: string) => {
+    return users.some((userEntry: User) => userEntry.user === userName);
+  };
+
+  const handleRemoveUser = async (userName: string) => {
+    const userAccess: userAccess = { name: userName, access: null };
+
+    await setModelSharingPermissions(
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:105]: A number of these arguments never change. The only ones that do are the users access, name and, type. You could clean up calls to this by having a curried version of this function that had the previous values pre-bound.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 13]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
             })}
           </div>
+          <div className="aside-split-col">
+            <h4>Add new user</h4>
+            <Formik
+              initialValues={{
+                username: "",
+                accessLevel: "read",
+              }}
+              // @ts-ignore
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:234]: Can this be removed?

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 14]
```diff
@@ -39,28 +124,232 @@ export default function ShareModel() {
...
+              validate={(values) => {
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.tsx:235]: Can you move these functions out of the markup plz.

**Review by hatched — CHANGES_REQUESTED** [review]
Thanks for this, I have added a number of comments but they are mostly just cleanup.

**Comment by hatched** [comment]
Design issues:

- The `Owner` cannot have it's access level  changed.
- The divider line doesn't go the full length of the panel if the number of users is short.
- The boxes in the left column are too large.
- Share icon and label in the header the wrong colour/size.
- "Add new user" the wrong size.

Screen Shot 2021-05-26 at 10 35 42 AM

File: .stylelintrc [hunk 15]
```diff
@@ -8,6 +8,7 @@
     "at-rule-no-unknown": [ true, {
       "ignoreAtRules": ["extend", "mixin", "include", "function", "if", "warn", "return", "at-root"]
     } ],
-    "no-invalid-position-at-import-rule": null
+    "no-invalid-position-at-import-rule": null,
+    "declaration-colon-newline-after": null
```
[the remark this answers is not in the document]
Comment (barrymcgee) [reply .stylelintrc:12]: Removing it doesn't break the linter so removed.

File: src/juju/index.js [hunk 16]
```diff
@@ -450,8 +452,59 @@ export async function executeActionOnUnits(
...
+          access,
+          action,
+          "model-tag": `model-${modelUUID}`,
+          "user-tag": `user-${user.name}`,
+        },
+      ],
+    });
+  };
+
+  if (previousAccess) {
```
[the remark this answers is not in the document]
Comment (barrymcgee) [reply src/juju/index.js:497]: Is it a performance overhead to revoke even if not needed?

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 3 above)
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:210]: Yes, the API returns any error from the grant/revoke calls here.

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 4 above)
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:127]: This is a temporary solution using Vanilla notifications until we have the toast noifications. To make it error, try to share a username that you're already sharing with.

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 8 above)
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:180]: Ahh.. these are a copy over from the actions forms which I forgot to delete. This also explains why I thought the FormikFormData component used in Actions was leaking into this view.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 17]
```diff
@@ -1,22 +1,89 @@
...
 import PanelHeader from "components/PanelHeader/PanelHeader";
+import FormikFormData from "components/FormikFormData/FormikFormData";
 
 import type { EntityDetailsRoute } from "components/Routes/Routes";
 import type { TSFixMe } from "types";
 
 import "./share-model.scss";
 
+type modelControllerData = {
+  additionalController: undefined;
```
[the remark this answers is not in the document]
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:22]: Screenshot 2021-05-27 at 12 11 17

File: src/components/Aside/_aside.scss (hunk 2 above)
Comment (barrymcgee) [reply src/components/Aside/_aside.scss:23]: The config panel is also split in this way... smells like a pattern to me.

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 7 above)
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:102]: If and when this function call moves out of scope, this change can be made.

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 12 above)
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:105]: I think it's clearer this way... there's already enough Russian dolls in this data flow without adding another one.

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (hatched) [reply src/panels/ShareModelPanel/ShareModel.tsx:21]: That means that the value is currently undefined, not that the type is. The type in this case is a `boolean`.

File: src/juju/index.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (hatched) [reply src/juju/index.js:497]: Yeah it might be, it's fine to leave it as-is.

File: src/components/Aside/_aside.scss (hunk 2 above)
Comment (hatched) [reply src/components/Aside/_aside.scss:23]: It is, but it collapses totally different. The columns do not stack on the config but they do here.

File: src/panels/ShareModelPanel/ShareModel.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx]: Done

File: src/components/Aside/_aside.scss (hunk 2 above)
Comment (barrymcgee) [reply src/components/Aside/_aside.scss:23]: Not a show stopper here

**Review by hatched — APPROVED** [review]
